plan/trace.go

```diff
@@ -0,0 +1,39 @@
+package plan
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/mysql"
+)
+
+// Trace represents a trace plan.
+type Trace struct {
+	baseSchemaProducer
+
+	StmtPlan Plan
+}
+
+// buildTrace builds a trace plan. Inside this method, it first optimize the
+// underlying query and then constructs a schema, which will be used to constructs
+// rows result.
+func (b *planBuilder) buildTrace(trace *ast.TraceStmt) (Plan, error) {
+	if _, ok := trace.Stmt.(*ast.SelectStmt); !ok {
+		return nil, errors.New("trace only supports select query")
+	}
+
+	optimizedP, err := Optimize(b.ctx, trace.Stmt, b.is)
+	if err != nil {
+		return nil, errors.New("fail to optimize during build trace")
```

Comment:
Why not return the error directly?

Reply:
creating a new error with error message `"fail to optimize during build trace"` seems to make a lot of senses.